### PR TITLE
Add ARM64 support to repo2docker

### LINKED ISSUE #1120: Unable to build locally "Could not open '/lib64/ld-linux-x86-64.so.2'"
### Bug description

On my computer I am unable to use `repo2docker` to build a local directory, even if it's empty. I am not sure why.

#### Expected behaviour

I expected `jupyter-repo2docker .` to build locally.

#### Actual behaviour
It fails:

```bash
$ jupyter-repo2docker .
Picked Local content provider.
Using local repo ..
Using PythonBuildPack builder
Step 1/48 : FROM buildpack-deps:bionic
 ---> 98fc26e1509f
Step 2/48 : ENV DEBIAN_FRONTEND=noninteractive
 ---> Using cache
 ---> 1ddee175e87f
Step 3/48 : RUN apt-get -qq update &&     apt-get -qq install --yes --no-install-recommends locales > /dev/null &&     apt-get -qq purge &&     apt-get -qq clean &&     rm -rf /var/lib/apt/lists/*
 ---> Using cache
 ---> 4888afea1faf
Step 4/48 : RUN echo "en_US.UTF-8 UTF-8" > /etc/locale.gen &&     locale-gen
 ---> Using cache
 ---> 560885b9842a
Step 5/48 : ENV LC_ALL en_US.UTF-8
 ---> Using cache
 ---> b20252556587
Step 6/48 : ENV LANG en_US.UTF-8
 ---> Using cache
 ---> 555a9faba040
Step 7/48 : ENV LANGUAGE en_US.UTF-8
 ---> Using cache
 ---> bac9774badef
Step 8/48 : ENV SHELL /bin/bash
 ---> Using cache
 ---> c32aa3e88981
Step 9/48 : ARG NB_USER
 ---> Using cache
 ---> ebc588c6509a
Step 10/48 : ARG NB_UID
 ---> Using cache
 ---> 0d37f8730abc
Step 11/48 : ENV USER ${NB_USER}
 ---> Using cache
 ---> fea8c4d34e03
Step 12/48 : ENV HOME /home/${NB_USER}
 ---> Using cache
 ---> ff2316ba9655
Step 13/48 : RUN groupadd         --gid ${NB_UID}         ${NB_USER} &&     useradd         --comment "Default user"         --create-home         --gid ${NB_UID}         --no-log-init         --shell /bin/bash         --uid ${NB_UID}         ${NB_USER}
 ---> Using cache
 ---> 504d5ac77f2a
Step 14/48 : RUN wget --quiet -O - https://deb.nodesource.com/gpgkey/nodesource.gpg.key |  apt-key add - &&     DISTRO="bionic" &&     echo "deb https://deb.nodesource.com/node_14.x $DISTRO main" >> /etc/apt/sources.list.d/nodesource.list &&     echo "deb-src https://deb.nodesource.com/node_14.x $DISTRO main" >> /etc/apt/sources.list.d/nodesource.list
 ---> Using cache
 ---> d994f49c133d
Step 15/48 : RUN apt-get -qq update &&     apt-get -qq install --yes --no-install-recommends        less        nodejs        unzip        > /dev/null &&     apt-get -qq purge &&     apt-get -qq clean &&     rm -rf /var/lib/apt/lists/*
 ---> Using cache
 ---> 9eb32056c343
Step 16/48 : EXPOSE 8888
 ---> Using cache
 ---> ca77d6d4c24a
Step 17/48 : ENV APP_BASE /srv
 ---> Using cache
 ---> 20995168e0d1
Step 18/48 : ENV NPM_DIR ${APP_BASE}/npm
 ---> Using cache
 ---> 788da69757a7
Step 19/48 : ENV NPM_CONFIG_GLOBALCONFIG ${NPM_DIR}/npmrc
 ---> Using cache
 ---> a5a1601fee12
Step 20/48 : ENV CONDA_DIR ${APP_BASE}/conda
 ---> Using cache
 ---> 5f2c97f735a5
Step 21/48 : ENV NB_PYTHON_PREFIX ${CONDA_DIR}/envs/notebook
 ---> Using cache
 ---> 28b20d442c48
Step 22/48 : ENV NB_ENVIRONMENT_FILE /tmp/env/environment.lock
 ---> Using cache
 ---> e72759ea9337
Step 23/48 : ENV KERNEL_PYTHON_PREFIX ${NB_PYTHON_PREFIX}
 ---> Using cache
 ---> 3adaed635bb7
Step 24/48 : ENV PATH ${NB_PYTHON_PREFIX}/bin:${CONDA_DIR}/bin:${NPM_DIR}/bin:${PATH}
 ---> Using cache
 ---> 1e8f85a6301f
Step 25/48 : COPY --chown=501:501 build_script_files/-2fopt-2fhomebrew-2fcaskroom-2fminiconda-2fbase-2flib-2fpython3-2e8-2fsite-2dpackages-2frepo2docker-2fbuildpacks-2fconda-2factivate-2dconda-2esh-91d161 /etc/profile.d/activate-conda.sh
 ---> Using cache
 ---> ab3cb148fe1e
Step 26/48 : COPY --chown=501:501 build_script_files/-2fopt-2fhomebrew-2fcaskroom-2fminiconda-2fbase-2flib-2fpython3-2e8-2fsite-2dpackages-2frepo2docker-2fbuildpacks-2fconda-2fenvironment-2epy-2d3-2e7-2elock-d00ccd /tmp/env/environment.lock
 ---> Using cache
 ---> 6ab76d080211
Step 27/48 : COPY --chown=501:501 build_script_files/-2fopt-2fhomebrew-2fcaskroom-2fminiconda-2fbase-2flib-2fpython3-2e8-2fsite-2dpackages-2frepo2docker-2fbuildpacks-2fconda-2finstall-2dminiforge-2ebash-ec0cb5 /tmp/install-miniforge.bash
 ---> Using cache
 ---> 047aef216f89
Step 28/48 : RUN mkdir -p ${NPM_DIR} && chown -R ${NB_USER}:${NB_USER} ${NPM_DIR}
 ---> Using cache
 ---> 4ce6e6ed3739
Step 29/48 : USER ${NB_USER}
 ---> Using cache
 ---> f7b89ca9c16b
Step 30/48 : RUN npm config --global set prefix ${NPM_DIR}
 ---> Using cache
 ---> 5f91a5ac3cff
Step 31/48 : USER root
 ---> Using cache
 ---> 88e5735d912c
Step 32/48 : RUN TIMEFORMAT='time: %3R' bash -c 'time /tmp/install-miniforge.bash' && rm -rf /tmp/install-miniforge.bash /tmp/env
 ---> Running in a9e4741c585a
++ dirname /tmp/install-miniforge.bash
+ cd /tmp
+ MINIFORGE_VERSION=4.9.2-2
+ MAMBA_VERSION=0.7.4
+ SHA256SUM=7a7bfaff87680298304a97ba69bcf92f66c810995a7155a2918b99fafb8ca1dc
+ URL=https://github.com/conda-forge/miniforge/releases/download/4.9.2-2/Mambaforge-4.9.2-2-Linux-x86_64.sh
+ INSTALLER_PATH=/tmp/miniforge-installer.sh
+ unset HOME
+ wget --quiet https://github.com/conda-forge/miniforge/releases/download/4.9.2-2/Mambaforge-4.9.2-2-Linux-x86_64.sh -O /tmp/miniforge-installer.sh
time: 12.992
+ chmod +x /tmp/miniforge-installer.sh
+ echo '7a7bfaff87680298304a97ba69bcf92f66c810995a7155a2918b99fafb8ca1dc  /tmp/miniforge-installer.sh'
+ sha256sum --quiet -c -
+ bash /tmp/miniforge-installer.sh -b -p /srv/conda
PREFIX=/srv/conda
Unpacking payload ...
qemu-x86_64: Could not open '/lib64/ld-linux-x86-64.so.2': No such file or directory
qemu-x86_64: Could not open '/lib64/ld-linux-x86-64.so.2': No such file or directory
time: 0.257
time: 13.550
Removing intermediate container a9e4741c585a
The command '/bin/sh -c TIMEFORMAT='time: %3R' bash -c 'time /tmp/install-miniforge.bash' && rm -rf /tmp/install-miniforge.bash /tmp/env' returned a non-zero code: 1
```
### How to reproduce
Create a new directory and run `repo2docker` on it.

### Your personal set up

 - OS: macOS 12.1 (21C52) Apple M1 Pro
 - Docker version: `docker version` 20.10.12
 - repo2docker version `repo2docker --version` 2021.08.0

```bash
$ docker version
Client:
 Cloud integration: v1.0.22
 Version:           20.10.12
 API version:       1.41
 Go version:        go1.16.12
 Git commit:        e91ed57
 Built:             Mon Dec 13 11:46:56 2021
 OS/Arch:           darwin/arm64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.12
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.16.12
  Git commit:       459d0df
  Built:            Mon Dec 13 11:43:07 2021
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.4.12
  GitCommit:        7b11cfaabd73bb80907dd23182b9347b4245eb5d
 runc:
  Version:          1.0.2
  GitCommit:        v1.0.2-0-g52b36a2
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### PR DESCRIPTION
**Pull request by yuvipanda**
I got a M1 Mac, what else does one do to procrastinate?

Fixes #1120 

TODO:
- [ ] Have frozen versions of the conda-lock files be architecture specific
- [ ] Be platform specific in getting R 
- [ ] Setup arm tests for everything

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -101,6 +101,12 @@ jobs:
           restore-keys: |
             ${{ runner.os }}-pip-
 
+      # Temporarily setup QEMU so we can run aarch64 build tests
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v2
+        with:
+          platforms: 'arm64,arm'
```
Comment (minrk) [inline .github/workflows/test.yml:108]: Running all the tests under qemu is going to take a _long_ time.

**Comment by yuvipanda** [comment]
I'm guessing this should be commandline param passed in.

File: .github/workflows/test.yml (hunk 1 above)
Comment (consideRatio) [reply .github/workflows/test.yml:108]: For reference: in z2jh we do the following, where `--platform` is passed during `docker build` is run by chartpress.

https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/d16d44aa44d43a3ac34a8fdec4a94328e47e2546/.github/workflows/test-docker-build.yaml#L47-L57

---

@yuvipanda I think you should only specify arm64 and not arm64,arm - at least unless you wish to go for arm32 support as well. Typically, the support for arm64 is quite more common, and in z2jh for example we provide arm64 + amd64 but not arm.

Comment (yuvipanda) [reply .github/workflows/test.yml]: @consideRatio yay ty. Just switched it to arm64.

@minrk I'm trying to get it to at least run on arm64 now, and can't get it to run at all :( I think it's still just running on amd64 despite me passing platform, not sure why.

**Comment by yuvipanda** [comment]
I see:

```
Warning: repo2docker:app.py:814  ---> [Warning] The requested image's platform (linux/arm64/v8) does not match the detected host platform (linux/amd64) and no specific platform was requested
```

everywhere, and am not sure if qemu is actually even being used? We don't use buildx, as we talk to it via the docker API (not `buildx`). Is it actually not possible to do this via the API? All the docs I can find seem to suggest using `buildx`

**Comment by manics** [comment]
Looks like docker-py doesn't support buildx https://github.com/docker/docker-py/issues/2230

**Comment by minrk** [comment]
> everywhere, and am not sure if qemu is actually even being used?

If the build steps don't fail immediately, qemu's involved. You'll get some kind of `exec error` if there's no qemu. It is weird that it thinks no platform is requested, though.

**Comment by yuvipanda** [comment]
@minrk they currently fail with:

```
2022-08-26T10:38:55.0109683Z INFO     repo2docker:app.py:814 [91m/tmp/install-base-env.bash: /srv/conda/envs/kernel/bin/ipython: /srv/conda/envs/kernel/bin/python: bad interpreter: No such file or directory
```

which I'd assume means they aren't using QEMU, and that's the arm binary trying to get started in an amd64 context.

It works on my local machine tho

**Comment by minrk** [comment]
Given how long our tests take, it may make sense to run arm tests on circle. I just did this on pyzmq, and it was ~10x faster than qemu on GHA.

Presumably we only need a pretty basic couple of tests, though.

**Comment by yuvipanda** [comment]
I dunno when I'll be able to work on this again though, would love for someone else to take it over <3

**Comment by yuvipanda** [comment]
A tarot card reader on the streets of los angeles told me that completion of this PR is not something she sees in my future given that Saturn is in retrograde, and I will hence close this in favor of the amazing #1228

**Pull request closed without merging**